File: docker/runners/python/runner.py
import json
import importlib.util
import sys
import traceback
from pathlib import Path

def load_function(module_path: Path, fn_name: str):
    spec = importlib.util.spec_from_file_location("user_code", module_path)
    mod = importlib.util.module_from_spec(spec)
    sys.modules["user_code"] = mod
    assert spec.loader
    spec.loader.exec_module(mod)
    return getattr(mod, fn_name)


def main():
    workdir = Path("/home/runner/app")
    tests_path = workdir / "tests.json"
    code_path = workdir / "user_code.py"
    result = {"tests": [], "error": None}
    try:
        data = json.loads(tests_path.read_text(encoding="utf-8"))
        fn_name = data.get("function_name")
        cases = data.get("tests", [])
        fn = load_function(code_path, fn_name)
        for idx, case in enumerate(cases):
            inputs = case.get("input", [])
            try:
                out = fn(*inputs)
                result["tests"].append({"name": case.get("name", f"test_{idx+1}"), "passed": out == case.get("expected"), "output": out, "expected": case.get("expected")})
            except Exception as exc:
                result["tests"].append({"name": case.get("name", f"test_{idx+1}"), "passed": False, "error": str(exc), "trace": traceback.format_exc()})
    except Exception as exc:
        result["error"] = str(exc)
    print(json.dumps(result, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: docker/runners/javascript/runner.js
const fs = require("fs");
const path = require("path");

function loadUserModule(codePath) {
  const mod = require(codePath);
  return mod;
}

function main() {
  const workdir = "/home/runner/app";
  const testsPath = path.join(workdir, "tests.json");
  const codePath = path.join(workdir, "user_code.js");
  const result = { tests: [], error: null };
  try {
    const data = JSON.parse(fs.readFileSync(testsPath, "utf-8"));
    const fnName = data.function_name;
    const cases = data.tests || [];
    const mod = loadUserModule(codePath);
    const fn = mod[fnName];
    if (typeof fn !== "function") throw new Error("Function not found: " + fnName);
    cases.forEach((c, idx) => {
      try {
        const out = fn.apply(null, c.input || []);
        result.tests.push({
          name: c.name || `test_${idx + 1}`,
          passed: JSON.stringify(out) === JSON.stringify(c.expected),
          output: out,
          expected: c.expected,
        });
      } catch (err) {
        result.tests.push({
          name: c.name || `test_${idx + 1}`,
          passed: false,
          error: err.message,
          trace: err.stack,
        });
      }
    });
  } catch (err) {
    result.error = err.message;
  }
  console.log(JSON.stringify(result));
}

main();
